compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -611,6 +664,20 @@ impl<'brillig> Context<'brillig> {
     }
 }
 
+fn instruction_can_be_hoisted(
+    instruction: &Instruction,
+    dfg: &mut DataFlowGraph,
+    deduplicate_with_predicate: bool,
+) -> bool {
+    // These two can never be hoisted as they have a side-effect
+    // (though it's fine to de-duplicate them, just not fine to hoist them)
+    if matches!(instruction, Instruction::Constrain(..) | Instruction::RangeCheck { .. }) {
+        return false;
+    }
+
+    instruction.can_be_deduplicated(dfg, deduplicate_with_predicate)
```

Comment:
This is implicit in the fact that we found a cached result: if the instruction couldn't be deduplicated it [would not be cached](https://github.com/noir-lang/noir/blob/7b28c597498fdbb16f336c3eea2c68ea9cfbb148/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L458).

Do you think this could be simplified to just: don't hoist `Instruction::Constrain`, which is the only instruction that feeds into the simplification mappings?

Reply:
@TomAFrench I think the pattern we see is something like this:
```rust
fn main(x: Field, y: bool, z: bool) {
    if y {
        assert_eq(x, 1);
    } else if z {
        assert_eq(x, 1);
    } 
    // Here we don't state anything about x, but because it appears in both branches above
    // the constraint was hoisted above the `if` and then fails any time x isn't 1.
}
```

Do you see any other instruction that can cause issues? I'd think all other [assertion intrinsics](https://github.com/noir-lang/noir/blob/7b28c597498fdbb16f336c3eea2c68ea9cfbb148/compiler/noirc_evaluator/src/ssa/ir/instruction.rs#L117-L130) are suspect to this, aren't they? So with https://github.com/noir-lang/noir/pull/6615 in mind, we should be calling `has_side_effect`, not `can_be_deduplicated(deduplicate_with_predicate)`. 

I'm a bit confused by the `use_predicate` thing because for example `Constrain` doesn't interact with the predicate and will use `None` as a key. It sounds like should be able to signal that we cannot use an expression under a different side-effect-variable, but it's not stored there under any, even though it can clearly depend on the `if`.

Reply:
I agree, we shouldn't hoist any instructions with side-effects out of their block as otherwise we can break ordering of side effects so we should switch to `has_side_effect`.

Reply:
Let me remind myself of `deduplicate_with_predicate`.